mo_net/model/layer/output.py

```
@@ -248,3 +261,320 @@ def output_dimensions(self) -> Dimensions:
+        Y_true: jnp.ndarray,
+    ) -> D[Activations]:
+        """Compute backward pass - sparse gradients only for paths to target words.
+
+        Args:
+            Y_true: Target word indices, shape (batch_size,)
+
+        Returns:
+            Gradient w.r.t. input activations, shape (batch_size, embedding_dim)
+        """
+        if (input_activations := self._cache["input_activations"]) is None:
+            raise ValueError("Input activations not set during forward pass.")
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)